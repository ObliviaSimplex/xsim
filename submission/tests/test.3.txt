=-=-=-=-=-=-=-=-=-=-=-=-=
Test 3: Signed Integers
=-=-=-=-=-=-=-=-=-=-=-=-=

This was a fairly basic set of tests, aiming to learn more about how xcpu
handles signed integer calculations. No bugs in the xcpu.c code were dis-
covered, but some strange features regarding the cmp instruction -- which
*could* be considered bugs, though my xcpu.c conformed with the gold ver-
sion on all outputs -- did turn up. Cmp, it turns out, *is* able to deliver
mathematically correct results for

-2 < -1        (0xfffe < 0xffff)
 1 <  2        (0x0001 < 0x0002)

and more generally, for

a < b --> b < a

*provided that b and a have the same sign*. When, for example, a is negative
and b is positive, cmp will not deliver mathemtically correct results, since
it treats every bitarray as representing an *unsigned* integer (and so gives
the perfectly reasonable response that 65535, say, is much greater than 1,
since this is how it interprets 0xffff and 0x0001 -- not as -1 and 1).

It would not, however, be difficult to provide a workaround for signed
comparisons. The snippet

mov      r1,  r3
mov      r2,  r4
load     0x8000, r0      # 0x8000 = 0b1000000000000000
and      r0,  r3
and      r0,  r4
equ      r3,  r4

Will tell us if two integers have the same sign. If they have the same sign,
then cmp will tell us their relative order. If they do not, then the negative
term is lower. 
