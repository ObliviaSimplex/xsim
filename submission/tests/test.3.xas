# -=-=-=-=-=-=-=-=-=-=-=-=-=-=-
# Test 3: Signed integers
# =-=-=-=-=-=-=-=-=-=-=-=-=-=-=

.glob main

main:
        std
        push    r14             # boilerplate:
        mov     r15,  r14       # saving the base pointer and stack pointer
        loadi   1,    r0
        loadi   2,    r1
        sub     r1,   r0        # 1-2
        cmp     r1,   r0        # this will give wonky results, due to cmp's
        cmp     r0,   r1        # undefined behaviour for signed integers
        mul     r0,   r1
        mul     r0,   r1
        mul     r0,   r1
        mul     r1,   r1
        cmp     r1,   r0
        cmp     r0,   r1
        sub     r0,   r1
        cmp     r1,   r0
        loadi   0xaaaa, r2
        loadi   0x5555, r3
        cmp     r3,   r2
        neg     r2
        neg     r3
        cmp     r2,   r3
        xor     r4,   r4
        xor     r5,   r5
        dec     r4
        dec     r4
        dec     r5
        cmp     r4,   r5
        neg     r4
        neg     r5
        cmp     r5,   r4        # a < b --> -b < -a holds, when signs same
        loadi   0xffff, r6
        mov     r6,   r7
        neg     r7
        equ     r6,   r7        
        xor     r0,   r0
        mov     r14,  r15
        pop     r14
        