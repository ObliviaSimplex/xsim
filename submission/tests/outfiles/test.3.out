PC: 040c, State: 0002: Registers:
 0000 0000 0000 0000 0000 0000 0000 0000 0000 0000 0000 0000 0000 0000 0000 0408
PC: 040e, State: 0002: Registers:
 0000 0000 0000 0000 0000 0000 0000 0000 0000 0000 0000 0000 0000 0000 0000 0406
PC: 0410, State: 0002: Registers:
 0000 0000 0000 0000 0000 0000 0000 0000 0000 0000 0000 0000 0000 0000 0406 0406
PC: 0414, State: 0002: Registers:
 0001 0000 0000 0000 0000 0000 0000 0000 0000 0000 0000 0000 0000 0000 0406 0406
PC: 0418, State: 0002: Registers:
 0001 0002 0000 0000 0000 0000 0000 0000 0000 0000 0000 0000 0000 0000 0406 0406
PC: 041a, State: 0002: Registers:
 ffff 0002 0000 0000 0000 0000 0000 0000 0000 0000 0000 0000 0000 0000 0406 0406
PC: 041c, State: 0003: Registers:
 ffff 0002 0000 0000 0000 0000 0000 0000 0000 0000 0000 0000 0000 0000 0406 0406
PC: 041e, State: 0002: Registers:
 ffff 0002 0000 0000 0000 0000 0000 0000 0000 0000 0000 0000 0000 0000 0406 0406
PC: 0420, State: 0002: Registers:
 ffff fffe 0000 0000 0000 0000 0000 0000 0000 0000 0000 0000 0000 0000 0406 0406
PC: 0422, State: 0002: Registers:
 ffff 0002 0000 0000 0000 0000 0000 0000 0000 0000 0000 0000 0000 0000 0406 0406
PC: 0424, State: 0002: Registers:
 ffff fffe 0000 0000 0000 0000 0000 0000 0000 0000 0000 0000 0000 0000 0406 0406
PC: 0426, State: 0002: Registers:
 ffff 0004 0000 0000 0000 0000 0000 0000 0000 0000 0000 0000 0000 0000 0406 0406
PC: 0428, State: 0003: Registers:
 ffff 0004 0000 0000 0000 0000 0000 0000 0000 0000 0000 0000 0000 0000 0406 0406
PC: 042a, State: 0002: Registers:
 ffff 0004 0000 0000 0000 0000 0000 0000 0000 0000 0000 0000 0000 0000 0406 0406
PC: 042c, State: 0002: Registers:
 ffff 0005 0000 0000 0000 0000 0000 0000 0000 0000 0000 0000 0000 0000 0406 0406
PC: 042e, State: 0003: Registers:
 ffff 0005 0000 0000 0000 0000 0000 0000 0000 0000 0000 0000 0000 0000 0406 0406
PC: 0432, State: 0003: Registers:
 ffff 0005 aaaa 0000 0000 0000 0000 0000 0000 0000 0000 0000 0000 0000 0406 0406
PC: 0436, State: 0003: Registers:
 ffff 0005 aaaa 5555 0000 0000 0000 0000 0000 0000 0000 0000 0000 0000 0406 0406
PC: 0438, State: 0003: Registers:
 ffff 0005 aaaa 5555 0000 0000 0000 0000 0000 0000 0000 0000 0000 0000 0406 0406
PC: 043a, State: 0003: Registers:
 ffff 0005 5556 5555 0000 0000 0000 0000 0000 0000 0000 0000 0000 0000 0406 0406
PC: 043c, State: 0003: Registers:
 ffff 0005 5556 aaab 0000 0000 0000 0000 0000 0000 0000 0000 0000 0000 0406 0406
PC: 043e, State: 0003: Registers:
 ffff 0005 5556 aaab 0000 0000 0000 0000 0000 0000 0000 0000 0000 0000 0406 0406
PC: 0440, State: 0003: Registers:
 ffff 0005 5556 aaab 0000 0000 0000 0000 0000 0000 0000 0000 0000 0000 0406 0406
PC: 0442, State: 0003: Registers:
 ffff 0005 5556 aaab 0000 0000 0000 0000 0000 0000 0000 0000 0000 0000 0406 0406
PC: 0444, State: 0003: Registers:
 ffff 0005 5556 aaab ffff 0000 0000 0000 0000 0000 0000 0000 0000 0000 0406 0406
PC: 0446, State: 0003: Registers:
 ffff 0005 5556 aaab fffe 0000 0000 0000 0000 0000 0000 0000 0000 0000 0406 0406
PC: 0448, State: 0003: Registers:
 ffff 0005 5556 aaab fffe ffff 0000 0000 0000 0000 0000 0000 0000 0000 0406 0406
PC: 044a, State: 0003: Registers:
 ffff 0005 5556 aaab fffe ffff 0000 0000 0000 0000 0000 0000 0000 0000 0406 0406
PC: 044c, State: 0003: Registers:
 ffff 0005 5556 aaab 0002 ffff 0000 0000 0000 0000 0000 0000 0000 0000 0406 0406
PC: 044e, State: 0003: Registers:
 ffff 0005 5556 aaab 0002 0001 0000 0000 0000 0000 0000 0000 0000 0000 0406 0406
PC: 0450, State: 0003: Registers:
 ffff 0005 5556 aaab 0002 0001 0000 0000 0000 0000 0000 0000 0000 0000 0406 0406
PC: 0454, State: 0003: Registers:
 ffff 0005 5556 aaab 0002 0001 ffff 0000 0000 0000 0000 0000 0000 0000 0406 0406
PC: 0456, State: 0003: Registers:
 ffff 0005 5556 aaab 0002 0001 ffff ffff 0000 0000 0000 0000 0000 0000 0406 0406
PC: 0458, State: 0003: Registers:
 ffff 0005 5556 aaab 0002 0001 ffff 0001 0000 0000 0000 0000 0000 0000 0406 0406
PC: 045a, State: 0002: Registers:
 ffff 0005 5556 aaab 0002 0001 ffff 0001 0000 0000 0000 0000 0000 0000 0406 0406
PC: 045c, State: 0002: Registers:
 0000 0005 5556 aaab 0002 0001 ffff 0001 0000 0000 0000 0000 0000 0000 0406 0406
PC: 045e, State: 0002: Registers:
 0000 0005 5556 aaab 0002 0001 ffff 0001 0000 0000 0000 0000 0000 0000 0406 0406
PC: 0460, State: 0002: Registers:
 0000 0005 5556 aaab 0002 0001 ffff 0001 0000 0000 0000 0000 0000 0000 0000 0408
CPU has halted.
